### PR TITLE
[sil-combine] Fix a subtle bug in getConformanceAndConcreteType related to handling of inherited conformances

### PR DESCRIPTION
**Pull request by swiftix**
This fixes a couple of tests and some of the compatibility suite projects.
rdar://problem/31815540, rdar://problem/31838976

It is supposed to be a proper fix for a problem addressed by a hack in #9038

### CODE REVIEW

**Comment by swiftix** [comment]
@slavapestov Do you mind having a look? The fix turned out to be even easier than expected.

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@DougGregor In theory, you should be able to revert your hack in #9038

**Comment by swiftix** [comment]
@slavapestov Hmm. There seems to be a failure in `tryDevirtualizeWitnessMethod`. The log is [here](https://ci.swift.org/job/swift-PR-osx-smoke-test/7439/testReport/junit/Swift(macosx-x86_64)/SILOptimizer/devirt_default_witness_method_sil/). The problem seems to be in  `getWitnessMethodSubstitutions`, the function that you and @DougGregor heavily modified.

**Comment by slavapestov** [comment]
Ok, this makes sense, but I would still like it if you made the other change where we use a SubstituionMap to build the new substitutions instead of directly manipulating the SubsitutionList. I would like to make SubstitutionList opaque soon and this is one of the few remaining places where we were relying on its structure.

**Comment by slavapestov** [comment]
@swiftix maybe the devirtualizer has a similar problem and the assertion found a legitimate bug there?

**Comment by swiftix** [comment]
@slavapestov Yes, I think it is a real problem probably. That's why I asked you to see how to fix it.

As for the SubstitutionList manipulation, I'll add the fix we had.

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 1]
```diff
@@ -686,27 +687,39 @@ SILCombiner::createApplyWithConcreteType(FullApplySite AI,
...
+            ProtocolType *proto) -> Optional<ProtocolConformanceRef> {
+          if (substTy->getCanonicalType() == ConcreteType) {
+            if (proto->getDecl() != Conformance.getRequirement()) {
+              assert(
+                  Conformance.getRequirement()->inheritsFrom(proto->getDecl()));
+              if (Conformance.isAbstract())
+                return ProtocolConformanceRef(proto->getDecl());
+              Conformance = ProtocolConformanceRef(
+                  Conformance.getConcrete()->getInheritedConformance(
+                      proto->getDecl()));
```
Comment (slavapestov) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: This bit and the similar snippet on line 813 could be factors out into a ProtocolConformanceRef::getInheritedConformance

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 2]
```diff
@@ -686,27 +687,39 @@ SILCombiner::createApplyWithConcreteType(FullApplySite AI,
...
+        FnTy->getGenericSignature()->getSubstitutionMap(AI.getSubstitutions());
+    auto FinalSubsMap = FnSubsMap.subst(
+        [&](SubstitutableType *type) -> Type {
+          if (type == OpenedArchetype)
+            return ConcreteType;
+          return type;
+        },
+        [&](CanType origTy, Type substTy,
+            ProtocolType *proto) -> Optional<ProtocolConformanceRef> {
+          if (substTy->getCanonicalType() == ConcreteType) {
```
Comment (slavapestov) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: You can use isEqual instead of getCanonicalType() ==

**Comment by swiftix** [comment]
@slavapestov Thanks for the suggestions. The code looks cleaner now.

Will you look into this `getWitnessMethodSubstitutions` issue?

**Comment by slavapestov** [comment]
Sure, but it will have to wait for tomorrow.

**Comment by slavapestov** [comment]
@swiftix You might want to check in the SIL combiner part of this patch and leave out the new assertion you added in GenericSignature::getSubstitutionMap(). When I fix devirt I will add the new assertion.

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Pull request merged**